src/scene/world-clusters.js

```diff
@@ -0,0 +1,663 @@
+        this._clusterBoundsDeltaData[1] = boundsDelta.y;
+        this._clusterBoundsDeltaData[2] = boundsDelta.z;
+
+        this._clusterCompressionLimit0Data[0] = this._maxAttenuation;
+        this._clusterCompressionLimit0Data[1] = this._maxColorValue;
+
+        // assign values
+        this._clusterPixelsPerCellId.setValue(this._pixelsPerCellCount);
+        this._lightsTextureInvSizeId.setValue(this._lightsTextureInvSizeData);
+        this._clusterTextureSizeId.setValue(this._clusterTextureSizeData);
+        this._clusterBoundsMinId.setValue(this._clusterBoundsMinData);
+        this._clusterBoundsDeltaId.setValue(this._clusterBoundsDeltaData);
+        this._clusterCellsDotId.setValue(this._clusterCellsDotData);
+        this._clusterCellsMaxId.setValue(this._clusterCellsMaxData);
+        this._clusterCompressionLimit0Id.setValue(this._clusterCompressionLimit0Data);
+    }
+
+    // evaluates min and max coordinates of AABB of the light in the cell space
+    evalLightCellMinMax(clusteredLight, min, max) {
+
+        // min point of AABB in cell space
+        min.copy(clusteredLight.min);
+        min.sub(this.boundsMin);
+        min.div(this.boundsDelta);
+        min.mul2(min, this.cells);
+        min.floor();
+
+        // max point of AABB in cell space
+        max.copy(clusteredLight.max);
+        max.sub(this.boundsMin);
+        max.div(this.boundsDelta);
+        max.mul2(max, this.cells);
+        max.ceil();
+
+        // clamp to limits
+        min.max(Vec3.ZERO);
+        max.min(this._cellsLimit);
+    }
+
+    collectLights(lights) {
+
+        // skip index 0 as that is used for unused light
+        const usedLights = this._usedLights;
+        let lightIndex = 1;
+
+        for (let i = 0; i < lights.length; i++) {
+
+            // use enabled and visible lights
+            const light = lights[i];
+            if (light.enabled && light.type !== LIGHTTYPE_DIRECTIONAL && light.visibleThisFrame) {
+
+                // within light limit
+                if (lightIndex < this.maxLights) {
+
+                    // reuse allocated spot
+                    let clusteredLight;
+                    if (lightIndex < usedLights.length) {
+                        clusteredLight = usedLights[lightIndex];
+                    } else {
+                        // allocate new spot
+                        clusteredLight = new ClusterLight();
+                        usedLights.push(clusteredLight);
+                    }
+
+                    // store light properties
+                    clusteredLight.light = light;
+                    light.getBoundingBox(tempBox);
+                    clusteredLight.min.copy(tempBox.getMin());
+                    clusteredLight.max.copy(tempBox.getMax());
+
+                    lightIndex++;
+                } else {
+                    console.warn("Clustered lighting: more than " + (this.maxLights - 1) + " lights in the frame, ignoring some.");
```

Comment:
should the console.warn be wrapped in _DEBUG?

Reply:
most of them are in the engine, so that's consistent

Reply:
yes most (all?) console logs are wrapped in _DEBUG, but this one isn't?